test/e2e/upgrade/alert/alert.go

```diff
@@ -0,0 +1,227 @@
+package alert
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/prometheus/common/model"
+
+	v1 "k8s.io/api/core/v1"
+	kapierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	watchtools "k8s.io/client-go/tools/watch"
+	"k8s.io/kubernetes/pkg/client/conditions"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+)
+
+const (
+	// Delay after upgrade is complete before checking for critical alerts
+	alertCheckSleep = 5 * time.Minute
+
+	// Previous period in which to check for critical alerts
+	alertPeriodCheckMinutes = 1
+
+	// Number of attempts if prometheus query fails
+	maxPrometheusQueryAttempts = 5
+
+	// Deley between each query attempt
+	prometheusQueryRetrySleep = 10 * time.Second
+)
+
+// UpgradeTest runs post-upgrade after alertCheckSleep delay and tests if any critical alerts are firing.
+type UpgradeTest struct {
+	url         string
+	bearerToken string
+	oc          *exutil.CLI
+}
+
+func (UpgradeTest) Name() string { return "check-for-critical-alerts" }
+func (UpgradeTest) DisplayName() string {
+	return "Check if critical alerts are firing after upgrade success"
+}
+
+// Setup creates parameters to query Prometheus
+func (t *UpgradeTest) Setup(f *framework.Framework) {
+	g.By("Setting up post-upgrade alert test")
+
+	// Must use version that's can run within Ginkgo It
+	oc := exutil.NewCLIWithoutNamespaceInside("prometheus")
+
+	var ok bool
+	url, bearerToken, ok := locatePrometheus(oc)
+	if !ok {
+		framework.Failf("Prometheus could not be located on this cluster, failing test %s", t.Name())
+	}
+	t.url = url
+	t.bearerToken = bearerToken
+	t.oc = oc
+	framework.Logf("Post-upgrade alert test setup complete")
+}
+
+// Test checks if any critical alerts are firing.
+func (t *UpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	g.By("Checking for critical alerts")
+
+	// Recover current test if it fails so test suite can complete
+	defer g.GinkgoRecover()
+
+	// Block until upgrade is done
+	g.By("Waiting for upgrade to finish before checking for critical alerts")
+	<-done
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Additonal delay after upgrade completion
+	g.By("Waiting before checking for critical alerts")
+	time.Sleep(alertCheckSleep)
+	cancel()
+
+	if len(os.Getenv("TEST_UNSUPPORTED_ALLOW_VERSION_SKEW")) > 0 {
+		e2eskipper.Skipf("Test is disabled to allow cluster components to have different versions, and skewed versions trigger multiple other alerts")
```

Comment:
Instead of duplicating logic like this for both the update and non-update suites, can we factor the logic out into helpers that are shared between the two consumers?

Reply:
Yes, agree. I was waiting until we sorted out "our side" before coming up with reusables and I also wasn't sure if we wanted to head down that path or just get our stuff out there. Now I know :)